File: learnLeetCode/Day51/Part3/quantum_nim.py
import sys

# Memoization table
memo = {}

def get_affected_indices(N, edges):
    adj = [[] for _ in range(N)]
    for u, v in edges:
        adj[u].append(v)
        adj[v].append(u)
    
    # Precompute masks or lists of affected indices for each pile move
    # Move on pile i affects i and all neighbors
    affected = []
    for i in range(N):
        indices = [i] + adj[i]
        affected.append(indices)
    return affected

def solve_quantum_nim(stones, edges):
    N = len(stones)
    affected = get_affected_indices(N, edges)
    
    # Clear memo for new game
    global memo
    memo = {}
    
    return can_win(tuple(stones), N, affected)

def can_win(state, N, affected):
    if state in memo:
        return memo[state]
    
    # Try all possible moves
    # A move: choose pile i, remove k stones
    # Valid if all piles in affected[i] have >= k stones
    
    # Optimization:
    # Iterate through piles.
    # For each pile i, max possible k is min(state[j] for j in affected[i])
    # If max_k == 0, no move possible on this pile.
    
    possible_moves = False
    
    for i in range(N):
        # Find max k
        max_k = state[i]
        for neighbor in affected[i]:
            if neighbor != i: # affected[i] includes i
                max_k = min(max_k, state[neighbor])
        
        if max_k == 0:
            continue
            
        # Try all k from 1 to max_k
        # Heuristic: Try largest k first? Or smallest?
        # Usually small k leads to more states.
        for k in range(1, max_k + 1):
            possible_moves = True
            
            # Construct next state
            next_state_list = list(state)
            for idx in affected[i]:
                next_state_list[idx] -= k
            next_state = tuple(next_state_list)
            
            if not can_win(next_state, N, affected):
                memo[state] = True
                return True
                
    memo[state] = False
    return False

if __name__ == "__main__":
    # Example 1: 3 piles, Line 0-1-2
    # Stones: [2, 2, 2]
    # Move 0: affects 0, 1.
    # Move 1: affects 0, 1, 2.
    # Move 2: affects 1, 2.
    
    stones = [3, 3, 3]
    edges = [(0, 1), (1, 2)]
    
    print(f"Stones: {stones}, Edges: {edges}")
    if solve_quantum_nim(stones, edges):
        print("Result: First Player Wins")
    else:
        print("Result: Second Player Wins")
        
    # Example 2: Disconnected
    stones2 = [2, 2]
    edges2 = []
    # Equivalent to Nim with piles 2, 2 -> XOR sum 0 -> Second wins.
    print(f"Stones: {stones2}, Edges: {edges2}")
    if solve_quantum_nim(stones2, edges2):
        print("Result: First Player Wins")
    else:
        print("Result: Second Player Wins")

File: Day51/Part3/quantum_nim.cpp
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>
#include <numeric>

using namespace std;

map<vector<int>, bool> memo;
vector<vector<int>> affected;
int N;

bool can_win(vector<int>& state) {
    if (memo.count(state)) return memo[state];
    
    // Try all moves
    for (int i = 0; i < N; ++i) {
        // Find max k
        int max_k = state[i];
        for (int neighbor : affected[i]) {
            if (neighbor != i) {
                max_k = min(max_k, state[neighbor]);
            }
        }
        
        if (max_k == 0) continue;
        
        for (int k = 1; k <= max_k; ++k) {
            vector<int> next_state = state;
            for (int idx : affected[i]) {
                next_state[idx] -= k;
            }
            
            if (!can_win(next_state)) {
                return memo[state] = true;
            }
        }
    }
    
    return memo[state] = false;
}

bool solve_quantum_nim(vector<int> stones, vector<pair<int, int>> edges) {
    N = stones.size();
    affected.assign(N, vector<int>());
    
    // Build adjacency
    vector<vector<int>> adj(N);
    for (auto& edge : edges) {
        adj[edge.first].push_back(edge.second);
        adj[edge.second].push_back(edge.first);
    }
    
    for (int i = 0; i < N; ++i) {
        affected[i].push_back(i);
        for (int neighbor : adj[i]) {
            affected[i].push_back(neighbor);
        }
    }
    
    memo.clear();
    return can_win(stones);
}

int main() {
    // Example 1
    vector<int> stones1 = {3, 3, 3};
    vector<pair<int, int>> edges1 = {{0, 1}, {1, 2}};
    
    cout << "Stones: {3, 3, 3}, Edges: {{0, 1}, {1, 2}}" << endl;
    if (solve_quantum_nim(stones1, edges1)) {
        cout << "Result: First Player Wins" << endl;
    } else {
        cout << "Result: Second Player Wins" << endl;
    }
    
    // Example 2
    vector<int> stones2 = {2, 2};
    vector<pair<int, int>> edges2 = {};
    
    cout << "Stones: {2, 2}, Edges: {}" << endl;
    if (solve_quantum_nim(stones2, edges2)) {
        cout << "Result: First Player Wins" << endl;
    } else {
        cout << "Result: Second Player Wins" << endl;
    }
    
    return 0;
}
